### DISCUSSION TITLE
Practical Fast Retrieval on Filecoin

### DISCUSSION THREAD

**Discussion by zixuanzh** [filecoin-project/FIPs#988] [Enhancements - Technical]
## **Abstract**

Enable Fast Retrieval on Filecoin by creating a second type of proof, dubbed "Fast PoRep", potentially with fewer layers of SDR (or other mechanisms) and requiring higher per-unit-storage initial pledge requirements to offset the loss in hardware security.

## **Motivation**

- One of the biggest challenges for the Filecoin Network to deliver value is that it is expensive to retrieve the data stored on Filecoin. 
    - There are L2s building on Filecoin as Hot Storage Layers but 1) GTM and fundraising can be quite hard 2) if Filecoin Retrieval itself is still very slow (fastest unsealing time requires around 10 minutes), then the argument to use Filecoin as a backend/value accrual can be weak.
  
- While the ecosystem can spend millions of dollars and months of research time to develop a new PoRep, it is also a design choice of the Filecoin protocol to prioritize hardware security. 

- As such, as a naive proposal, it is worth considering to reduce the hardware security from the number of layers of SDR required in the current PoRep (or other forms of proof) to enable fast sealing and unsealing quickly. 

- This then creates a marketplace of PoRep with a tradeoff over their security and usability. 

- To counterbalance the loss in security over usability with fewer number of SDR layers or other forms of proofs, the per unit-storage initial pledge requirement of such new PoRep should also increase.

## **Design Rationale/Consideration**

- Time to market is key. 
    - which includes research and implementation time
    - only if we also believe this FIP moves the needle for Filecoin

- The goal is to find a solution that can be live **by the next upgrade**, hence "practical", making Filecoin more usable, lowering hardware requirements, and locking more tokens to compensate for the loss in hardware security.

- As such, this FIP tries to avoid shaking existing business models on the network and create a positive sum outcome in the ecosystem. 

- SPs can choose different PoReps that suit their needs (similar FoFR regardless of which PoRep that SPs choose from a mining perspective). One can argue for a lower FoFR for Fast PoRep since there is less hardware commitment and require higher security compensation from initial pledge, locking up more Filecoin as usage increases.

**Comment by nicola** [comment]
This is a good direction. Let me give you some context on why specifically lowering the number of layers is not enough:
* Reducing the number of layers of SDR will make PoRep not secure. PoRep needs some number of layers - what we have today - in order to mantain the security of space-hardness needed for consensus security. In other words, if you remove the layers you lose security both in the cost model and in the time model, since there is no way to force an SP to reconstruct the sector if they were to delete more than 20% (the spacegap). Another way to view this is that you are increasing the spacegap - which would make PoRep not suitable for storage-backed consensus security (and if you are ok losing that, we have a new upcoming proposal - happy to collaborate - where we give up storage-backed security and at that point you can use a simpler Proof of Data Possession, which much simpler and requires no sealing).
* Reducing the number of layers will not really give "fast retrieval", the fact that the PoRep layers are sequential will play a big penalty on your latency - e.g. if you were to go to half the layers will still take 30-45 mins on a CPU and way more on GPUs

**Reply by zixuanzh** [reply]
It doesn't have to be reducing the number of PoRep layers in the current PoRep. It's just a naive hypothetical solution. Goal is to find something that works quickly (time is of the essence), without disrupting the current business models (to reduce blocker from social consensus), and create a positive sum outcome. Would love to collaborate on Proof of Data Possession and introduce it on L1 to enable fast retrieval soon.

**Reply by lucaniz** [reply]
hey @zixuanzh what nicola is saying is that having fast retrieval by modifying PoRep is not a fast path. It is not by eliminating some layers (which in general does not work  as nicola pointed out) that you get fast retrieval possiblity.

We are planning to get back on Proof of data possession/ proof of access very soon (writing a doc on it). 
Let me stress this would make fast retrieval POSSIBLE via clear copy of the data, checking  periodical access to the data itself. 
Note that this is not a guaranteed retriavl as the SP can always refuse to serve the data. 
But imho is a fast retrieval enablement, which is a big step forward wrt status quo.

**Reply by jnthnvctr** [reply]
@lucaniz - I'd also love to read when its shareable! Related if you have a view - what would be involved (are these like relatively easy to implement things, or require many quarters of r&d / eng). 

I think one of the good points that @zixuanzh brings up is the time-premium making fast access commands - would be good to figure out what folks would be able to do to help accelerate. 

Has anyone done any looking into other "practical"-ish solutions for fast retrieval? 

From my understanding the rough landscape is: 
- (current) Extra clear copy (as an optional unsealed thing) => basically relying on handshake agreements, off-chain legal contracts perhaps
- I think @laudiacay had once suggested something using Lurk precompiles in the FEVM to build sort of Sia-like protocol for fast access copies? Perhaps could couple w/ retriev?
- I'm actually not sure if Claudia's suggestion was super close / far from @ribasushi 's suggestion which involved I think doing a PoST over a second copy (in the clear)?
- (New porep? [PoL based stuff](https://filecoinproject.slack.com/archives/C04USV0L7PF/p1713022425495669)? seems like this is all in the theory realm)

**Reply by lucaniz** [reply]
@jnthnvctr let me take this opportunity to give an overview of past Cryptonet work (which is gonna be the starting point of future work on this topic).

Some time ago, I proposed (with @irenegia ) to provably enable fast retrieval capabilities by introducing extra copy checks. 
The idea was that in order to make a first significant step towards fast retrievability, we had to first provably enable that that was possible. By challenging a clear copy of the data you have such guarantee: the SP has indeed periodical access to (at least) a copy of the data, thus he can serve them fast if needed (and of course if he is willing to serve data to its clients).
Note that this solution (which afaiu is really similar to the one @laudiacay  was proposing recently) inherits the fact that extra copy check is an independent routine from windowpost/winningpost, thus needing its own scheduling and accounting. 
Back then those particular needs were flagged as problematic, thus we never really progress beyond the "idea phase".
Anyone interested can find the overall idea [here ](https://cryptonet.org/checking-clear-data-in-filecoin). 
Note that in this doc we link also to older work Irene and I did when working on new PoRep which basically says that if you want retrieval in terms of milliseconds, the easiest way to get there is via extra copy. The doc (put together by @irenegia  summarizing the broader work on PoRep) can be found [here](https://cryptonet.org/blog/product-problem-fast-retrieval).

Some months later, while discussing with @ribasushi about the "independent scheduling/accounting" issue, he had the intution of getting implicit access to clear data while embedding this check in the current WindowPost, WinningPost schedule, thus keeping extra copy check entangled with consensus. 

I formalized the intuition and spec out the details [here](https://cryptonet.org/notebook/fil-poa-filecoin-proof-of-access). Note that there is also a FAQ section on FIL+ interaction. The TL;DR here is that if at any WindowPost adn WinningPoST SP prove both on replica and sector key, they implicitly prove access to a clear copy of the data. Note that one can pass the proving step by simply storing the replica and having access to a (potentially shared) copy of the data in the clear, since Replica = SectorKey + Data [modular sum] thus not incurring in major storage overhead. The main feature of this approach is you can use the same schedule as today for proving and there is a straightforward power accounting mechanism (now you lose power if you miss any WindowPost on replica (apart the first one), with this you'd lose power if you miss either the WIndowPOst on the replica or on the SectorKey, as they are proven together).

Now, the two approaches are basically equivalent in terms of guarantees they provide on fast retrieval enablement, but they are slightly different in terms of their features/how they interact with the current protocol (as I sketched above). Deciding which path to take is the most urgent thing imho (and of course feedbacks are more than welcome). 

That said, I stress that we should align in what all above (and this holds also for ZX idea/proposal). Everything we are talking about is enabling fast retrieval capability. It is NOT proving fast retrievability. 
In other words, we would be proving that fast retrievability is possible (which is a big step forward imho) but not that it is granted. Indeed, a malicious provider can always pass the "provable access step" and then refure to serve the data (at his discretion). 

If we want to have provable retrievability we basically need a trusted third party of some form. Note that in this directions we proposed in the past two different approaches: 

- Retrieve: project led by @irenegia, which you can learn more [here](https://cryptonet.org/projects/data-retrievability) which is basically a committee based solution to resolve retrievability disputes
-  Storage Metrics DAO, which aimed to be a decentralized reliable reputation system for storage-based features checking (you can find more [here ](https://www.notion.so/pl-strflt/Storage-Metrics-DAO-905d09995eea4c84842183348aec45e5))

**Reply by deltazxm** [reply]
What is the next plan about Fast Retrieval ?

**Reply by momack2** [reply]
Current status is that FilOz is working on a draft FIP for the “proof of retrievable copy” Luca described above, aiming for hopeful community acceptance before EOQ 👍

**Reply by zixuanzh** [reply]
hey @lucaniz , do you need to store an extra copy of the data on top of the copy that you use for mining for either Proof of Data Possession (separate proof of clear data) and Proof of Access (WindowPoSt and WinningPoSt on sectorkey+replica)? Trying to understand do we double the “storage requirement” for either of these

**Reply by zixuanzh** [reply]
In addition, what's the hardware requirement like (if we can give a percentage estimate) for running those two proofs vs SDR today?

**Reply by lucaniz** [reply]
You'd have to have access of a clear copy of the data (meaning that this copy can be shared if the file is stored by multiple entities). Or you can use an extra copy of the data if for some reason you are already storing one for backup reasons

**Reply by lucaniz** [reply]
> In addition, what's the hardware requirement like (if we can give a percentage estimate) for running those two proofs vs SDR today?

you can think of the additional proving step as something comparable with WindowPoSt

**Comment by jaychoucc** [comment]
Fast retrieval is the most important issue for filecoin

**Comment by dd45e640b42e6da7da96faee3996ef7c** [comment]
does this proposal indicate that we can exchange [hardware security] and [PoS security] at a certain rate? 

> To counterbalance the loss in security over usability with fewer number of SDR layers or other forms of proofs, the per unit-storage initial pledge requirement of such new PoRep should also increase.


because then we can just PoS the L1 completely, by far the easiest, fastest, solution for the problems this proposal wants to work on and move all the annoying proof stuff to L2 market implementations.

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
> SPs can choose different PoReps that suit their needs (similar FoFR regardless of which PoRep that SPs choose from a mining perspective). One can argue for a lower FoFR for Fast PoRep since there is less hardware commitment and require higher security compensation from initial pledge, locking up more Filecoin as usage increases.

as soon as we offer 2+ ways to PoST on the L1 for blockrewards we have a competition between those 2+ solutions. the market will be fast to eliminate all but one of them from being used.

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
![2024-04-14_10-50](https://github.com/filecoin-project/FIPs/assets/60559388/653f3e28-64ca-48e3-8fd6-3f52fda2515a)

good example of the effects of offering 2+ solutions with vastly differing cost profiles to achieve the same outcome regarding FoFR

**Reply by zixuanzh** [reply]
I am trying to reduce scope of changes to reduce both r&d and consensus building time. I think just doing PoS on L1 is a huge change to Filecoin. Changing L1 consensus to PoS also does not solve the problem of enabling fast retrieval.

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
it allows to scale the cost of porep hw security to the storage clients needs on the spot without compromising chain security

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
> To counterbalance the loss in security over usability with fewer number of SDR layers or other forms of proofs, the per unit-storage initial pledge requirement of such new PoRep should also increase.

more important it solves the problem of years of lame duck governance. we can quit the dance around where to put the needle on the PoW - PoS meter. no one whatsoever benefited from SDM, CDM, fip36/56...78....80...stuff. no one! it is a colossal waste of resources and it leaves the project weaker than before. 

we managed to implement exactly 0 crypto economical changes since we decided on fip26, in 2021! i wouldn't even call this a crypto economic fip tbh. last real change was october 14th 2020 with fip04

you are smart in putting this out in an early stage to get it shut down or approved on a theoretical level. before r&d is done. because if this gets close/r to getting a community vote be sure i will drag you into another SDM/CDM situation to try to move the needle on the meter more to where i would like to see it. and as long as you put out proposals that devalue miners investments i can do this very efficiently and effective. not moving means stability...

> does this proposal indicate that we can exchange [hardware security] and [PoS security] at a certain rate?

if yes - there is no reason to not choose PoW or PoS. i honestly rather have PoS than more PoS - the direction all of your proposals want to move the network in. i am happy to listen to the argument for keeping both, always. keeping both has a proven track record - an ugly one - try your luck. there is no credible version of its future to sell anymore, sry - oversold.

it is how it is, isn't it? i do not enjoy this - at all. - believe me. nothing personal thou, just money!

**Comment by jennijuju** [comment]
> fastest unsealing time requires around 10 minutes)

is there any more info on how is this achieved, Whcih also means sealing can be done in 10min?

**Reply by ribasushi** [reply]
☝️ seconded. @zixuanzh if what you are saying: `unsealing time of 10 minute` does indeed exist - this is **really bad**. That would mean complete invalidation of the current Filecoin security model. Please share your information/findings either here or alternatively with `security@filecoin.org` as soon as possible.

**Reply by irenegia** [reply]
Yes,  "Sealing in 10 mins" could pose a security problem. To understand if that is the case, we need to know how actually this particular sealing procedure is done (we need details on HW used, storage used, etc)! So it would be great to know more and riba's suggestion to share info by email is good, but we can also do a call.

**Reply by ribasushi** [reply]
@jennijuju @cryptonemo @nicola can one of the protocol-design groups start some sort of incident doc on this or whatnot? This remark on the existence of a *"10 minute SDR"* should be thoroughly investigated. Initiatives like e.g. `PoA` hinge on the security of `SDR` being sound.

**Reply by momack2** [reply]
@zixuanzh do you mean that folks can unsnap in 10 mins? that'd be expected. But SDR sealing/unsealing is expected ~2.5hr minimum

**Comment by willscott** [comment]
I think this fip / proposal is somewhat mis-named. This is not a getting us to 'fast retrieval' but rather is a proof of unsealed copy.

The reasons we are aware of why why content is not made available in filecoin is both a risk and economic problem, and this will chip away at the problem, but it does not address the concept that people expect when they hear 'retrieval'. that needs to both include something like retriev that has a bounty to offset the potential liabilities of serving user generated content, and a protocol which makes the serving of data profitable, while also structuring expected SLAs such that the provider can meet them without expensive/un-needed over-provisioning.

**Reply by lucaniz** [reply]
FYI
> https://github.com/filecoin-project/FIPs/discussions/988#discussioncomment-9185872

**Reply by jennijuju** [reply]
Will - for proof of retrievable copy  I believe @lucaniz is planning to propose a **separate** FIP discussion and see https://github.com/filecoin-project/FIPs/discussions/988#discussioncomment-9165391, we agree on  clearly setting the expectation that we believe such new proof can help making fast retrieval POSSIBLE (no need to wait for unsealing).

**Reply by jennijuju** [reply]
And I agree with what you are raising. Should this fip discussion be retitled to “theoretical fast retrieval on filecoin” instead? 😆

**Reply by steven004** [reply]
I appreciate the term "a proof of unsealed copy."

**Reply by zixuanzh** [reply]
This is just a FIP discussion and there is no formal proposal yet. I wanted to align us on the requirements first. Ultimately, I think we want 1) something that we can go to market quickly, hence "practical" 2) make Filecoin L1 a strong dependency for L2s or other projects (without some ability for fast retrieval on L1, the dependency can be quite weak) 3) increase demand for using the FIL token
